### PR TITLE
Add replies collection

### PR DESCRIPTION
**Pull request by Menrath**
Add [replies collection](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-replies) to posts and federated comments.

**Solves Issue** https://github.com/Automattic/wordpress-activitypub/issues/138

### Description
The replies collection offer a neat way to sync replies over multiple instances. Mostly on demand and sometimes also client side on demand. They are not yet frequently used because of the moderation issues that arise with them. But since WordPress already got a good approve-system for comments this should not be too problematic here.

### Rest Endpoint
To make sure the ID's of the collection and the collections page are resolveable.
It also adds a new rest endpoint `/activitypub/1.0/(posts|comments)/<id>/replies` which returns a `Collection`. Adding the parameter `?page=0` returns the CollectionPage directly.
I decided not to use something like `<the_permalink>/replies` because it is only needed for `activity+json` and has no frontend view. Seemed consistent to me.

### Unit tests:
I started writing new unit tests, but was not sure how deep to test this. I'm happy for some discussions.

### Manual testing:
Write federated comments for a post, or receive comments and make sure those comments are approved.
When querying the ActivbityPub's JSON representation of the post the replies key exists, e.g.:

```json
...
 "replies": {
    "id": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies",
    "type": "Collection",
    "first": {
      "id": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=0",
      "type": "CollectionPage",
      "partOf": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies",
      "items": [
        "https://mz.lan/comments/8be36040-5a02-4335-917f-4c5dcc988e1f",
        "https://mz.lan/comments/de26c344-8d7d-400f-a8f5-5b89307bb72a",
        "https://mz.lan/comments/abce32af-a59e-41c6-8b47-a8925980f1ef",
        "https://wp.lan/?c=13",
        "https://mz.lan/comments/fa3d2ee2-147b-4eee-b682-98b0b7115603"
      ]
    }
  },
...
```

Changing the global option `comments_per_page` to for example "3" would change the response to this:

```json
...
"replies": {
    "id": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies",
    "type": "Collection",
    "first": {
      "id": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=0",
      "type": "CollectionPage",
      "partOf": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies",
      "items": [
        "https://mz.lan/comments/8be36040-5a02-4335-917f-4c5dcc988e1f",
        "https://mz.lan/comments/de26c344-8d7d-400f-a8f5-5b89307bb72a"
      ],
      "next": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=1"
    }
  },
...
```

Note that the items only contains 2 IDs. This is because the code currently uses core functions to query the paginated comments and later filters out the local-only ones. Because we have more pages also the key `next` is present.

It (`https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=1`) resolves to:

```json
**https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=1**
{
  "@context": [
    "https://www.w3.org/ns/activitystreams",
    {
      "Hashtag": "as:Hashtag"
    }
  ],
  "id": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=1",
  "type": "CollectionPage",
  "partOf": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies",
  "items": [
    "https://mz.lan/comments/abce32af-a59e-41c6-8b47-a8925980f1ef",
    "https://wp.lan/?c=13"
  ],
  "next": "https://wp.lan/wp-json/activitypub/1.0/posts/49/replies?page=2"
}
```

**Unapproved or local only comments should never be listed.**

### Discussion points
- This draft has always functions handle both replies for comments and posts rather than splitting up the code in the beginning. I just chose this construct to get to a working prototype and I already fear that there might be good arguments to do it differently.
- The code which gets the federated comment IDs: https://github.com/Menrath/wordpress-activitypub/blob/replies_collection/includes/collection/class-replies.php#L108 These lines are duplicated or have very similar occurences, more than once in the code.
- Maybe use a direct meta-query to get all approved-federated comments in a paginated way.
- I am happy to hear any ideas and thoughts.
- Mastodon for example lists all self-replies on the first page of the collection, and external replies later, this would also be possible, to ensure threads are displayed. Threads are possible now also in WordPress, but maybe not that important.
- Mastodon includes local replies fully, not only the ID. Within WordPress this in theory could get difficult because of the JSON-LD context (different post types with different ActivityPub object types at worst).

### CODE REVIEW

File: includes/collection/class-replies.php
Comment (pfefferle) [inline includes/collection/class-replies.php:14]: I think we should merge that class with the interactions collection. I have no emotions about which one to use, so we can wait until this is merged and migrate the interactions class afterwards. Just wanted to mention it.

I think maybe `Replies` is really the better choice!

**Comment by pfefferle** [comment]
Awesome 😍

**Comment by pfefferle** [comment]
@Menrath Is this ready to merge at least for a first attempt?

**Comment by Menrath** [comment]
I would love to see some more unit tests, but besides that I feel from my point of view I have nothing to add atm :)

File: includes/collection/class-replies.php
Comment (Menrath) [reply includes/collection/class-replies.php:14]: I prefer smaller files and think having this under collection/ is a good place.

File: includes/rest/class-collection.php
Comment (pfefferle) [inline includes/rest/class-collection.php:305]: You can use an `enum` for that, so that you do not have to check it in the action.

See: https://github.com/Automattic/wordpress-activitypub/blob/8bff92d26020c575782f50740dece4f6c6ed7247/includes/rest/class-followers.php#L134

File: includes/rest/class-collection.php
Comment (pfefferle) [inline includes/rest/class-collection.php:307]:
```suggestion
			'type' => 'string',
			'enum' => array( 'post', 'comment' ), 
```

File: includes/collection/class-replies.php
Comment (pfefferle) [reply includes/collection/class-replies.php:14]: I agree that this should be a collection and I will think about the "small" files :)

File: includes/collection/class-replies.php
Comment (pfefferle) [inline includes/collection/class-replies.php:81]: couldn't this be included in `get_collection_page` with `$page = 0`???

Comment (pfefferle) [reply includes/collection/class-replies.php:81]: @Menrath friendly ping ☺️

**Review by Menrath — COMMENTED** [review]
Seems all good to me.

File: includes/collection/class-replies.php
Comment (Menrath) [reply includes/collection/class-replies.php:117]: @pfefferle Should someone have a look at this before this PR gets merged? I could find time in the next days.

Comment (pfefferle) [reply includes/collection/class-replies.php:117]: but be aware that there are two different usecases:

1. We want to have the ID with URL as fallback.
2. We want to have the URL with ID as fallback.

**Review by pfefferle — APPROVED** [review]

File: includes/collection/class-replies.php
Comment (Menrath) [reply includes/collection/class-replies.php]: https://github.com/Automattic/wordpress-activitypub/pull/876/commits/769e434f02b7acdbe641ca60c5488b5b7c90e64c added a static functions to the Comment util class, that tries to get the link via the comments meta.  At some points I thought better error-catching could be possible, but I think this is beyond the scope of this PR. We could also just merge the current state, and revert the commit and add another PR for refactoring the code for comment related stuff.

**Review by pfefferle — APPROVED** [review]

**Comment by pfefferle** [comment]
thanks a lot @Menrath !

**Comment by Menrath** [comment]
@pfefferle You're welcome! :) I also really like the last changes, splitting the functionality into the `get_source_url` and `get_source_id`. It's just easier to read.

**Pull request merged**
